src/platform/nrfconnect/OTAImageProcessorImpl.cpp

```diff
@@ -81,11 +90,41 @@ CHIP_ERROR OTAImageProcessorImpl::ProcessBlock(ByteSpan & block)
     VerifyOrReturnError(mDownloader != nullptr, CHIP_ERROR_INCORRECT_STATE);
 
     CHIP_ERROR error = ProcessHeader(block);
-
     if (error == CHIP_NO_ERROR)
     {
-        // DFU target library buffers data internally, so do not clone the block data.
-        error = System::MapErrorZephyr(dfu_target_write(block.data(), block.size()));
+        mCurrentImage.mCurrentOffset += block.size();
+        if (mCurrentImage.mCurrentOffset >= mCurrentImage.mFileInfo->mFileSize)
+        {
+            // calculate how many data should be moved to the next image
+            uint64_t remainingDataSize = mCurrentImage.mCurrentOffset - static_cast<uint64_t>(mCurrentImage.mFileInfo->mFileSize);
+            // write last data of previous image
+            error = System::MapErrorZephyr(dfu_target_write(block.data(), block.size() - remainingDataSize));
+            // switch to net image
+            mCurrentImage.mIndex++;
+            mCurrentImage.mFileInfo = &mContentHeader.mFiles[mCurrentImage.mIndex];
+            if (OTAImageContentHeader::FileId::kNetMcuboot == mCurrentImage.mFileInfo->mFileId &&
+                mCurrentImage.mFileInfo->mFileSize > 0 &&  CHIP_NO_ERROR == error)
+            {
+                ChipLogDetail(SoftwareUpdate, "Remaining bytes from last block containing Net core data: %" PRIu64, remainingDataSize);
+                dfu_target_init(DFU_TARGET_IMAGE_TYPE_MCUBOOT, mCurrentImage.mIndex, /* size */ 0, nullptr);
+                // write remaining data to new image
+                error =
+                    System::MapErrorZephyr(dfu_target_write(block.data() + (block.size() - remainingDataSize), remainingDataSize));
+                mCurrentImage.mCurrentOffset = remainingDataSize;
+            }
+            else 
+            {
+                // set an error to end download process if the second image is not a Net core image.
```

Comment:
This comment may not be accurate, e.g. when other condition fails like `error` is not `CHIP_NO_ERROR` etc.